src/main/java/com/project/serverpool/service/ServerService.java

```diff
@@ -0,0 +1,138 @@
+
+import java.util.*;
+
+@Service
+public class ServerService {
+
+    private final ServerDao serverDao = new ServerDao();
+    private final ClientDao clientDao = new ClientDao();
+
+    /*
+        hash(server):{
+                         "server": Server,
+                         "available capacity": 40,
+                         "clients wait to locate a recourse":List<Client>
+                     }
+    */
+    private Map<Integer, Map<String, Object>> waitingList = new HashMap<>();
+
+
+
+    public List<Server> getAllServerPools() {
+        return serverDao.getAll();
+    }
+
+    public  Server locateServer(Client client) throws Exception {
+        Server server = new Server();
+        boolean isRecourseLocatedInCreatingSate = false;
+        synchronized (server){
+            server = serverDao.findAvailableServer(client.getMemory());
+        }
+        clientDao.save(client);
+
+        //no available recourse in the current active servers
+        //check the servers in the creation state
+        if (server == null) {
+            //loop over waiting list array to check if is there available resource in the creating state servers
+            for (Map.Entry<Integer, Map<String, Object>> entry : waitingList.entrySet()) {
+                //ensure that only one thread is read/write the list
+                //to read the most recent data from the list
+                synchronized (waitingList){
+                    int serverKey = entry.getKey();
+                    Map<String, Object> serverData = entry.getValue();
+                    Server entryServer = (Server) serverData.get("server");
+                    //if the current server has available recourse
+                    if ((Long) serverData.get("avaCapacity") >= client.getMemory()) {
+                        long leftRecourse = (Long) serverData.get("avaCapacity") - client.getMemory();
+                        List<Client> clients = (List<Client>) serverData.get("clients");
+                        clients.add(client);
+                        entryServer.setCapacity(leftRecourse);
+                        entryServer.setClients(clients);
+                        serverData.put("avaCapacity", leftRecourse);
+                        serverData.put("clients", clients);
+                        serverData.put("server", entryServer);
+                        waitingList.put(serverKey, serverData);
+                        // #TODO:solve return issue
+                        //server in the creating state
+                        isRecourseLocatedInCreatingSate = true;
+                        server =entryServer;
+                    }
+                }
+            }
+            // current waiting list array do not have available resource in the creating state servers
+            // spin new server
+            if (isRecourseLocatedInCreatingSate){
+                //wait until finish creating
+                while (waitingList.get(server.hashCode()) != null){
```

Comment:
not a good practice, how its working with you ?

Reply:
when a client wants to locate a resource on the server and the server on the creating state (it will be on the waiting list )
I locate the resource from the waiting list and wait for the server to finish its configuration (waiting 20 sec) then return the server as a result
if I return the server on line 66, it will return the server on the creating state  

I tried to find a better way than this, but I did not figure how😅
any advice?